app/components/post_draft/draft_handler/draft_handler.tsx

```diff
@@ -0,0 +1,296 @@
+        }
+        setUploadError(error);
+
+        uploadErrorTimeout.current = setTimeout(() => {
+            setUploadError(null);
+        }, UPLOAD_ERROR_SHOW_INTERVAL);
+    }, []);
+
+    const addFiles = useCallback((newFiles: FileInfo[]) => {
+        if (!newFiles.length) {
+            return;
+        }
+
+        if (!canUploadFiles) {
+            newUpdateError(uploadDisabledWarning(intl));
+            return;
+        }
+
+        const availableCount = maxFileCount - currentFiles.length;
+        if (newFiles.length > availableCount) {
+            newUpdateError(fileMaxWarning(intl, maxFileCount));
+            return;
+        }
+
+        const largeFile = newFiles.find((file) => file.size > maxFileSize);
+        if (largeFile) {
+            newUpdateError(fileSizeWarning(intl, maxFileSize));
+            return;
+        }
+
+        for (const file of newFiles) {
+            prepareUpload(file, fileHandler, serverUrl, channelId);
+        }
+
+        setUploading(true);
+        newUpdateError(null);
+
+        const newFileList = [...currentFiles, ...newFiles];
+        updateFiles(newFileList);
+    }, [intl, maxFileCount, maxFileSize, serverUrl, channelId, currentFiles, updateFiles]);
+
+    const removeFile = useCallback((file: FileInfo) => {
+        const handler = fileHandler.current[file.clientId!];
+        if (handler) {
+            handler.cancel?.();
+        }
+
+        const newFileList = currentFiles.filter((f) => f.clientId !== file.clientId);
+        updateFiles(newFileList);
+    }, [currentFiles, updateFiles]);
+
+    const retryFileUpload = useCallback((file: FileInfo) => {
+        const index = currentFiles.findIndex((f) => f.clientId === file.clientId);
+        if (index == null) {
+            return;
+        }
+
+        const newFile = {...file};
+        newFile.loading = true;
+        newFile.progress = 0;
+        newFile.failed = false;
+
+        prepareUpload(newFile, fileHandler, serverUrl, channelId);
+        setUploading(true);
+        updateFiles([...currentFiles.slice(0, index), newFile, ...currentFiles.slice(index + 1)]);
+    }, [serverUrl, channelId, currentFiles, updateFiles]);
+
+    useEffect(() => {
+        return () => {
+            if (uploadInterval.current) {
+                clearInterval(uploadInterval.current);
+            }
+            if (uploadErrorTimeout.current) {
+                clearTimeout(uploadErrorTimeout.current);
+            }
+        };
+    }, []);
+
+    useEffect(() => {
+        if (uploading && !uploadInterval.current) {
+            uploadInterval.current = setInterval(() => handlerJob.current(), UPLOAD_INTERVAL);
+        }
+
+        if (!uploading && uploadInterval.current) {
+            clearTimeout(uploadInterval.current);
+            uploadInterval.current = undefined;
+        }
+    }, [uploading]);
+
+    useEffect(() => {
+        handlerJob.current = () => {
+            let changed = false;
```

Comment:
The handlerJob is being called inside an interval. If we were to have it as a normal function, a change in the props (currentFiles and updateFiles) will require us to stop the interval, update the function, and start the interval again.

I felt it makes things simpler and I don't remember if I had more pressing reasons to do it this way (performance maybe?), but I realize that it may be a bit obscure. If you prefer, I could try to go the other way around: on change stop the interval and start it again with the correct function.

Reply:
Ah mimicking the instance property by using the ref; that's fine :)